pycbc/events/stat.py

```diff
@@ -567,6 +568,109 @@ def single(self, trigs):
+        alphas = coeff_file['fit_coeff'][:]
+        rates = coeff_file['count_in_template'][:]/\
+                float(coeff_file.attrs['analysis_time'])
+        # the template_ids and fit coeffs are stored in an arbitrary order
+        # create new arrays in template_id order for easier recall
+        tid_sort = numpy.argsort(template_id)
+        return {'alpha':alphas[tid_sort], 'rate':rates[tid_sort],
+                'thresh':coeff_file.attrs['stat_threshold']}
+
+    def get_ref_vals(self, ifo):
+        self.alphamax[ifo] = self.fits_by_tid[ifo]['alpha'].max()
+
+    def find_fits(self, trigs):
+        """Get fit coeffs for a specific ifo and template id(s)"""
+        try:
+            tnum = trigs.template_num  # exists if accessed via coinc_findtrigs
+            ifo = trigs.ifo
+        except AttributeError:
+            tnum = trigs['template_id']  # exists for SingleDetTriggers
+            # Should only be one ifo fit file provided
+            assert len(self.ifos) == 1
+            ifo = self.ifos[0]
+        # fits_by_tid is a dictionary of dictionaries of arrays
+        # indexed by ifo / coefficient name / template_id
+        alphai = self.fits_by_tid[ifo]['alpha'][tnum]
+        ratei = self.fits_by_tid[ifo]['rate'][tnum]
+        thresh = self.fits_by_tid[ifo]['thresh']
+        return alphai, ratei, thresh
+
+    def lognoiseratedensity(self, trigs):
+        """
+        Calculate the log noise rate density over single-ifo newsnr
+        Read in single trigger information, make the newsnr statistic
+        and rescale by the fitted coefficients alpha and rate
+        """
+        alphai, ratei, thresh = self.find_fits(trigs)
+        newsnr = self.get_newsnr(trigs)
+        # alphai is constant of proportionality between single-ifo newsnr and
+        #  negative log noise likelihood in given template
+        # ratei is rate of trigs in given template compared to average
+        # thresh is stat threshold used in given ifo
+        lognoisel = numpy.array(numpy.log(alphai) + numpy.log(ratei)
+                                - alphai * (newsnr - thresh),
+                                ndmin=1, dtype=numpy.float32)
+        return lognoisel
+
+    def single(self, trigs):
+        """
+        Single-detector statistic, here just equal to the log noise
+        rate density
+        """
+        return self.lognoiseratedensity(trigs)
+
+    def coinc_multiifo(self, s, slide,
+                  step, time_addition=0.005): # pylint:disable=unused-argument
+        """Calculate the final coinc ranking statistic"""
+        ifostr = ' '.join(sorted(self.ifos))
+
+        ratesprod = sum(x for x in s.values())
+        ifo_list = [ifo for ifo in self.fits_by_tid]
+        ln_coinc_area = numpy.log(coinc_rate.multiifo_noise_coincident_area(
+                                                     ifo_list, time_addition))
+
+        loglr = mean_log_coincidence_rates[ifostr] - ln_coinc_area - ratesprod
```

Comment:
@GarethDaviesGW why are you normalizing by an ifo-dependent mean rate?

Reply:
The statistic we want is the rate of noise coincidences (as a density over stat value). The rate of noise coincidences is dependent on the ifo combination. That is the point of calculating the rates for each ifo combinations separately, they *are* different. I suggested scaling out by a 'normal/standard' H1L1 rate simply in order to make the numbers manageable.